kiwi/system/prepare.py

```diff
@@ -273,25 +274,31 @@ def install_system(self, manager):
                     'System archive installation failed: %s' % format(e)
                 )
 
-    def pinch_system(self, manager, force=False):
+    def pinch_system(self):
         """
-        Delete packages marked for deletion in the XML description
-
-        :param object manager: instance of a :class:`PackageManager` subclass
-        :param bool force: force deletion true|false
+        Delete packages marked for deletion in the XML description. First
+        uninstalls packages marked with `type="uninstall"` if any, then
+        proceeds with packages marked with `type="delete"` if any.
 
         :raises KiwiInstallPhaseFailed: if the deletion packages process fails
         """
         to_become_deleted_packages = \
             self.xml_state.get_to_become_deleted_packages()
-        if to_become_deleted_packages:
-            log.info('Pinch system')
-            try:
-                self.delete_packages(manager, to_become_deleted_packages, force)
-            except Exception as e:
-                raise KiwiInstallPhaseFailed(
-                    '%s: %s' % (type(e).__name__, format(e))
+        to_become_uninstalled_packages = \
+            self.xml_state.get_to_become_uninstalled_packages()
+        try:
+            if to_become_uninstalled_packages:
+                self.delete_packages_without_configured_repos(
+                    to_become_uninstalled_packages
+                )
```

Comment:
I would have thought this would be:

```python
self.delete_packages(manager, to_become_uninstalled_packages, force=False)
```

Reply:
We could do it this way, but then we need a manager instance in `build` and `prepare` tasks. It would be also ok to handle the new manager instance in `prepare` and `build` tasks before running `pinch_system`. This way the system_prepare API would not include this weirdness.

Reply:
The reason why this delete statement exists is to free the resources of package manager and repository instance bound to the manager instance such that the config.sh script code can not play tricks with it while we are building. So imho we have two options:

* move del manager after pinch_system, which would allow config.sh writers to influence the temp repo setup. Probably we don't want to open this gate

* update the pinch_system method to work if manager is not set

   ```
   def pinch_system(self, manager=None, force=False):
       if not manager:
            ...
   ```

Thoughts ?

Reply:
and then just change the pinch_system call to

```
system.pinch_system(force=...)
```

and prevent the additional delete_packages_with... method

Reply:
I do not consider the first option to be feasible, it opens a range of cases to consider, note that the `imageinclude` and `imageonly` repos are already in place for `config.sh` with the current setup, altering this could cause issues.

I am fine with altering pinch_system to work with an undefined manager. I have some doubts about `force` argument. I am fine with it, but then we would have two `pinch_system` calls together:

```python
# Remove type="uninstall" packages
system.pinch_system(manager=None, force=False)
# Remove type="delete" packages
system.pinch_system(manager=None, force=True)
```
I am fine with both options, (i) having a `pinch_system` method that does a single operation (with force argument) or (ii) having a `pinch_system` method that performs both (without force argument).